### PR TITLE
Add integration tests using k3d to run a Kubernetes cluster

### PR DESCRIPTION
**Pull request by colega**
This adds some base for writing integrations tests.

While not strictly needed yet, I do want to have integration tests to finish https://github.com/grafana/rollout-operator/pull/25 

The integration tests run in a Kubernetes cluster orchestrated by [k3d](https://github.com/k3d-io/k3d). Initially [I tried running a docker container containing k3d](https://github.com/grafana/rollout-operator/pull/28) but that required extra effort to communicate with k3d and it seemed just like too many indirection levels for me (think about it: _CI is a docker container that runs a test that runs a docker container that runs k3d that runs docker containers that run kubernetes nodes that run docker containers_).

So instead, I just impoted the `k3d` code. I took a look into the commands that create, delete and load images into a cluster and with all of them I build `integration/k3t` package (see what I did here? k3s is a small k8s, k3d is k3s in docker, now k3t is k3d for tests).

Most of the ~1600 files changed here are just the import of k3d and its dependencies. None of this code will actually reach production (well, except for the dependencies that we already had and we had to update).

You can see how `k3t.NewCluster()` creates a new cluster that is automatically destroyed on test cleanup, and it provides a `kubernetes.Clientset` that can be used to interact with that cluster. It also prints the `KUBECONFIG=` env var with the path of the KUBECONFIG used. Useful for local debugging.

I just wrote one happy case for the initial rollout, I already spent a lot of time on this task and want to stop here, but it's straightforward to prepare a second test that checks that `rollout-operator` doesn't rollout with unhealthy pods in a different zone. 

This also makes a change in the rollout-operator itself making the reconcileInterval configurable now, which is useful for making the tests run quicker.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -22,6 +22,16 @@ jobs:
...
       - run: make test
 
+  integration:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '^1.19.4'
+      - run: make rollout-operator build-image
```
Comment (pracucci) [inline .github/workflows/ci.yaml:32]: Why do you also need to make the `build-image` here?

File: cmd/rollout-operator/main.go [hunk 2]
```diff
@@ -24,18 +25,19 @@ import (
...
+	kubeConfigFile := flag.String("kubernetes.config-file", "", "The API config file path. If not specified, it will be auto-detected when running within a API cluster.")
+	kubeNamespace := flag.String("kubernetes.namespace", "", "The API namespace for which this operator is running.")
+	reconcileInterval := flag.Duration("reconcile.interval", 5*time.Second, "The interval of reconciliation.")
 	logLevel := flag.String("log.level", "debug", "The log level. Supported values: debug, info, warn, error.")
 	flag.Parse()
 
 	// Validate CLI flags.
 	if *kubeNamespace == "" {
-		fatal(errors.New("The Kubernetes namespace has not been specified."))
+		fatal(errors.New("The API namespace has not been specified."))
```
Comment (pracucci) [inline cmd/rollout-operator/main.go:37]: Is this change intentional? Same across the rest of the file, where doesn't look right to me.

File: integration/k3t/cluster.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	return c.k
+}
+
+func (c Cluster) LBPort() string {
+	return c.lbPort
+}
+
+func (c Cluster) LoadImages(ctx context.Context, t *testing.T, images ...string) {
+	t.Helper()
+	// TODO: we're using k3d.ImportModeTooldNode mode because the auto-detect (that uses pipe) fails in CI: https://github.com/k3d-io/k3d/issues/900
```
Comment (pracucci) [inline integration/k3t/cluster.go:128]: [nit]
```suggestion
	// TODO: we're using k3d.ImportModeToolsNode mode because the auto-detect (that uses pipe) fails in CI: https://github.com/k3d-io/k3d/issues/900
```

File: integration/util_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/grafana/rollout-operator/integration/k3t"
+)
+
+// requireEventuallyPod runs the tests provided on the pod retrieved from the API.
+// Once a test suceeds, it's not evaluated again.
```
Comment (pracucci) [inline integration/util_test.go:21]: [nit]
```suggestion
// Once a test succeeds, it's not evaluated again.
```

**Review by pracucci — APPROVED** [review]
Massive work, LGTM! I'm very happy to see integration tests for the rollout-operator. I left few minor comments (major is about the "Kubernetes" renaming to "API" which doesn't look right to me).

Please also remove the binary `integration/mock-service/test-service`.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (colega) [reply .github/workflows/ci.yaml:32]: I need the image for the integration tests. Maybe there's something obvious I'm missing?

File: cmd/rollout-operator/main.go (hunk 2 above)
Comment (colega) [reply cmd/rollout-operator/main.go]: Definitely not intentional, reverted in https://github.com/grafana/rollout-operator/pull/30/commits/b69aaa35866a9d73aea263908fd94678d487edce.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (pracucci) [reply .github/workflows/ci.yaml]: Update after an offline discussion. `build-image` builds the `rollout-operator` image and not a "build image" image (like in Mimir). I got confused by it. Maybe we don't need `make rollout-operator` tho.

Comment (colega) [reply .github/workflows/ci.yaml]: Removed `rollout-operator` target in https://github.com/grafana/rollout-operator/pull/30/commits/449a7d9871d118ab5189bec997881668ff090f66

File: integration/e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/grafana/rollout-operator/integration/k3t"
+)
+
+func TestRolloutHappyCase(t *testing.T) {
+	ctx := context.Background()
+
+	k3t.ConfigureLogger(t)
```
[the remark this answers is not in the document]
Comment (colega) [reply integration/e2e_test.go]: Done in https://github.com/grafana/rollout-operator/pull/30/commits/fc39b96b672e7b1ecf67ead818bbc9be995f1003, with an option to not doing it.

File: integration/e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	cluster := k3t.NewCluster(ctx, t, fmt.Sprintf("k3d-test-%d", time.Now().UnixMilli()))
```
[the remark this answers is not in the document]
Comment (colega) [reply integration/e2e_test.go]: Done in https://github.com/grafana/rollout-operator/pull/30/commits/fc39b96b672e7b1ecf67ead818bbc9be995f1003, with an option to provide it explicitly.

File: integration/util_test.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		if phase == expectedPhase {
+			t.Logf("Pod %q phase is the expected %q", pod.Name, phase)
+			return true
+		}
+		t.Logf("Pod %q phase %q is not the expected %q.", pod.Name, phase, expectedPhase)
+		return false
+	}
+}
+
+func expectReady() func(t *testing.T, pod *corev1.Pod) bool {
```
[the remark this answers is not in the document]
Comment (colega) [reply integration/util_test.go:52]: Done in https://github.com/grafana/rollout-operator/pull/30/commits/1cb4c73912501c7d075d2490c83c92a49fccbc7c

File: integration/util_test.go (hunk 4 above)
Comment (colega) [reply integration/util_test.go]: Done in https://github.com/grafana/rollout-operator/pull/30/commits/1cb4c73912501c7d075d2490c83c92a49fccbc7c

File: integration/k3t/cluster.go (hunk 3 above)
Comment (colega) [reply integration/k3t/cluster.go]: Done in https://github.com/grafana/rollout-operator/pull/30/commits/1cb4c73912501c7d075d2490c83c92a49fccbc7c

**Review by pracucci — APPROVED** [review]
Thanks for addressing my feedback. LGTM!
